smarts/sstudio/types.py

```diff
@@ -595,11 +600,320 @@ class Traffic:
+                    )
+                return result
+        elif self.inner_condition.evaluate(
+            *args, simulation_time=simulation_time, **kwargs
+        ):
+            object.__setattr__(self, key, simulation_time)
+        return ConditionState.FALSE
+
+    def __post_init__(self):
+        if self.inner_condition.__class__ in _abstract_conditions:
+            raise TypeError(
+                f"Abstract `{self.inner_condition.__class__.__name__}` cannot use delay operations."
+            )
+
+
+@dataclass(frozen=True)
+class OnRoadCondition(SubjectCondition):
+    """This condition is true if the subject is on road."""
+
+    def evaluate(self, *args, vehicle_state, **kwargs) -> ConditionState:
+        return ConditionState.TRUE if vehicle_state.on_road else ConditionState.FALSE
+
+
+@dataclass(frozen=True)
+class VehicleTypeCondition(SubjectCondition):
+    """This condition is true if the subject is of the given vehicle types."""
+
+    vehicle_type: str
+
+    def evaluate(self, *args, vehicle_state, **kwargs) -> ConditionState:
+        return (
+            ConditionState.TRUE
+            if vehicle_state.vehicle_config_type == self.vehicle_type
+            else ConditionState.FALSE
+        )
+
+
+@dataclass(frozen=True)
+class VehicleSpeedCondition(SubjectCondition):
+    """This condition is true if the subject has a speed between low and high."""
+
+    low: float
+    """The lowest speed allowed."""
+
+    high: float
+    """The highest speed allowed."""
+
+    def evaluate(self, *args, vehicle_state, **kwargs) -> ConditionState:
+        return (
+            ConditionState.TRUE
+            if self.low <= vehicle_state.speed <= self.high
+            else ConditionState.FALSE
+        )
+
+    @classmethod
+    def loitering(cls: Type["VehicleSpeedCondition"], abs_error=0.01):
+        return cls(low=abs_error, high=abs_error)
+
+
+@dataclass(frozen=True)
+class CompoundCondition(Condition):
+    """This compounds multiple conditions.
+
+    The following cases are notable
+        CONJUNCTION (A AND B)
+            If both conditions evaluate TRUE the result is exclusively TRUE.
+            Else if either condition evaluates EXPIRED the result will be EXPIRED.
+            Else if either condition evaluates BEFORE the result will be BEFORE.
+            Else FALSE
+        DISJUNCTION (A OR B)
+            If either condition evaluates TRUE the result is exclusively TRUE.
+            Else if either condition evaluates BEFORE then the result will be BEFORE.
+            Else if both conditions evaluate EXPIRED then the result will be EXPIRED.
+            Else FALSE
+        IMPLICATION (A AND B or not A)
+            If the first condition evaluates *not* TRUE the result is exclusively TRUE.
+            Else if the first condition evaluates TRUE and the second condition evaluates TRUE the result is exclusively TRUE.
+            Else FALSE
```

Comment:
This is how the condition operators evaluate.

Notable oddities are:
- On conjunction, if either condition is expired the conjunction will always be expired.
- On disjunction, if both conditions are expired the conjunction will always be expired.
- On implication, the resulting condition can only be TRUE or FALSE. The only theoretical case where BEFORE or EXPIRED can occur is when the first condition evaluates as a constant TRUE (via LiteralCondition and negations and combinations of) but I have not handled that case.